### ISSUE TITLE
GCC 14: abseil-src/absl/numeric/internal/bits.h:309:24: error: '__builtin_ctzs' needs isa option -mbmi

### ISSUE THREAD

**Issue by fruffy** [abseil/abseil-cpp#1664]
### Describe the issue

Our CI on Fedora 40, which ships GCC 14 now, has been hitting `abseil-src/absl/numeric/internal/bits.h:309:24: error: '__builtin_ctzs' needs isa option -mbmi` recently in the Abseil library. 
https://github.com/p4lang/p4c/actions/runs/8880814585/job/24381737348

We are not sure how to fix this problem (discussion at https://github.com/p4lang/p4c/pull/4638). Currently, we use `-march=native` but that seems like a big hammer. Do you by any further advice on how to resolve this problem? 

### Steps to reproduce the problem

Compile Abseil on Fedora 40 with GCC 14

### What version of Abseil are you using?

[20240116.1](https://github.com/abseil/abseil-cpp/releases/tag/20240116.1)

### What operating system and version are you using?

Fedora 40

### What compiler and version are you using?

gcc-14.0.1-0.15.fc40.x86_64 

### What build system are you using?

cmake-3.28.2-1.fc40.x86_64

**Comment by derekmauro** [comment]
It looks like we can use `__builtin_clzg` here.

**Comment by ChrisDodd** [comment]
Why not just use `__builtin_ctz` here?  Per gcc docs `__builtin_ctz`T returns an unspecified value for a 0 input, so `ctz` might return a well-defined value where `ctzs` would return unspecified (when the bottom 16 bits are all 0).  I guess that it would be desirable to use the more efficient 16-bit instruction from the bmi instruction set when that is available, but using the 32 bit instruction should be fine when it is not.  Overall, this really seems like a gcc bug -- `__builtin_`X should use whatever sequence of instructions is appropriate for any target.  Saying it is not supported when it is perfectly supportable seems wrong.

**Comment by derekmauro** [comment]
There is no reason to guess when you have benchmarks.

**Comment by pinskia** [comment]
>Overall, this really seems like a gcc bug -- __builtin_X should use whatever sequence of instructions is appropriate for any target. Saying it is not supported when it is perfectly supportable seems wrong.

Note __builtin_ctzs was never documented and it was only even an x86 specific builtin and was only supposed to be an internal to GCC's x86 intrinsics but folks started to use it anyways. And then clang decided to add it as a `generic` builtin too. 

It was originally removed from GCC 6.1.0 but had to be added back because folks used it incorrectly: See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=59874 for some of the history here.

Basically abseil was always wrong in using __builtin_ctzs in the first place. 

Second if you want a defined value at 0, you could just do `a == 0 ? value : __builtin_ctz(a);` and that will be optimized to just the instruction if that instruction specifies the value at 0. Or secondary for GCC 14+, you can use `__builtin_ctz(a, value)` instead.
